torch/fx/passes/fuser_utils.py

```diff
@@ -0,0 +1,228 @@
+        if indegree_map[node] == 0:
+            candidates.put(node)
+
+    sorted_nodes: NodeList = list()
+    while not candidates.empty():
+        node = candidates.get()
+        sorted_nodes.append(node)
+
+        for n in node.users:
+            if n in indegree_map:
+                indegree_map[n] -= 1
+                if indegree_map[n] == 0:
+                    candidates.put(n)
+
+    return sorted_nodes
+
+
+def validate_partition(partition: NodeList) -> bool:
+    # verify the partition does't form a dependency cycle in the original graph
+    # returns True for valid partition, False for invalid
+
+    partition_set = set(partition)
+
+    outputs: NodeList = list()
+    for node in partition_set:
+        for user_node in node.users:
+            if user_node not in partition_set:
+                # external user node, need to expose as an output
+                outputs.append(user_node)
+
+    # perform DFS on the parition outputs
+    # if it reaches a node within the partition, then it found a cycle
+    visited: NodeSet = set()
+    def dfs_find_cycle(node):
+        if node in partition_set:
+            return False  # found cycle, return False for invalid
+
+        visited.add(node)
+        for user_node in node.users:
+            if user_node not in visited:
+                if not dfs_find_cycle(user_node):
+                    return False
+        return True
+
+    for output_node in outputs:
+        if not dfs_find_cycle(output_node):
+            return False
+
+    return True
+
+
+# TODO: copied from split_by_tags's impl, refactor split_by_tags to use this function
+def copy_module_attributes(gm: torch.fx.GraphModule, subgraph: torch.fx.Graph) -> HolderModule:
+    # Loop through all module calls (call_module) and param fetches (get_attr)
+    # in this component, creating HolderModules as necessary to match the path.
+    # e.g. if in the original module there's a get_attr node fetches "conv.weight".
+    # We create a HolderModule as root -> add a HolderModule named "conv" ->
+    # make "weight" a attribute of "conv" HolderModule and point to conv.weight in
+    # the original module.
+    submodule = HolderModule({})
+    for n in subgraph.nodes:
+        if n.op not in ("call_module", "get_attr"):
+            continue
+
+        target = n.target
+        assert isinstance(target, str)
+        target_name_parts = target.split(".")
+        curr = submodule
+        orig_gm = gm
+
+        for name in target_name_parts[:-1]:
+            if not hasattr(curr, name):
+                curr.add_module(name, HolderModule({}))
+
+            curr = getattr(curr, name)
+            orig_gm = getattr(orig_gm, name)
+
+        leaf_node_name = target_name_parts[-1]
+        leaf_node = getattr(orig_gm, leaf_node_name)
+
+        # Relies on custom __setattr__ magic.
+        setattr(curr, leaf_node_name, leaf_node)
+
+    return submodule
+
+def fuse_partition(gm: torch.fx.GraphModule,
```

Comment:
Should this be implemented as a `transformer` pattern to keep the code shorter?

Reply:
IIUC, Transformer pattern is relative localized to a particular node. This function needs to work with a subgraph. 
It's probably ok to implement it with a Transformer, we can refactor this in the next iteration.